### DISCUSSION TITLE
More versatile led indicators

### DISCUSSION THREAD

**Discussion by ArtTales** [afpineda/OpenSourceSimWheelESP32#16] [Ideas]
It would be great if this project could display more data, such as drs availability, kers, time delta or different types of rpm indications (sequential, blocks, start showing lights before red zone, not from start) , maybe some indications when cluch engaged, launch assist lights. Or even better if it was possible to customise via simhub 
![image](https://github.com/user-attachments/assets/4b0f01ea-bebe-4d09-a040-ab1633d89b76)
Here's an example what i'm talking about: https://imgur.com/a/7QLLBhH
Thank you for your awesome work!

**Comment by afpineda** [comment]
> Or even better if it was possible to customise via simhub 

This was my desire, too. The best way to deal with LEDs is to give full control over them to SimHub. 
This is compatible with the "telemetry data" approach and may come in the future. 
However, my knowledge is limited. I found no way to create a custom device for SimHub. 
SimHub is a fantastic piece of work, but there is no documentation available for third-party developers.

This is all I can do for now, since I don't get any help from "the community".

**Reply by ArtTales** [reply]
Yeah, like you said, the sim racing community in developing something really sucks. If only I could help but I only have pretty basic programming knowledge. Maybe it’s worth creating some kind of community in discord, I think after some time there will be that person who will help. By the way, do you know anything about ffbbeast? Developer of this software is currently working on integrating Simhub into his product

**Comment by afpineda** [comment]
I did some research just to be sure about the LED control in SimHub. 
Here is the situation:

- SimHub relays in its own Arduino sketeches for LED control.
- The communication between SimHub and Arduino is via USB virtual serial port, **exclusively**.
- The communication protocol is public, but only works on USB serial devices.
- This project uses HID reports instead of serial connections, so there is no way to implement this protocol in the device itself.
- A workaround is to install a *null modem* (aka *virtual port emulator*) driver on the host PC, then redirect SimHub's serial data to a custom application. This application would translate the SimHub data into HID reports. SimHub expects a USB COM port with a particular VID and PID. So, this workaround may not work.

There is currently no suitable virtual port emulator available. There are some commercial virtual port emulators, but they are not free or open source, except for trial versions. 
The only free alternative is "com0com", which is outdated and does not work on Windows 11 due to security concerns.

**Comment by janc18** [comment]
I can help with that, and we can use the [Virtual HID Framework](https://learn.microsoft.com/es-es/windows-hardware/drivers/hid/virtual-hid-framework--vhf-) to create the virtual port. However, we would need to create an application or process to enable communication between the two devices.

**Reply by afpineda** [reply]
Thanks for the help, but don't go too fast. 
I'm not sure it will work.

Go to SimHub > Add device > Add standard device > Standard LED device.

We have to trick SimHub into thinking a port emulator is a USB serial connection to an Arduino board, 
but I don't know how it goes into port listing and device detection.

I will ask for more information in the official SimHub Discord server.

Edit: 
Discord conversation started here: [https://discordapp.com/channels/299259397060689920/453962746162315264/1307015534696468591](https://discordapp.com/channels/299259397060689920/453962746162315264/1307015534696468591)

**Comment by afpineda** [comment]
I will record the Discord conversation here (anonimously) for later reference:

> Hi. Help needed here. I'm developing an open source firmware for custom sim wheels/button boxes based on HID reports (no serial connection) both in BLE (wireless) and USB (wired). I already managed to send telemetry data to the device using a SimHub plugin. However, I need to send RGB LED color data to the custom device as well. In short, I need the custom device to work as a "Standard Led Device" (in SimHub) but not based in the Arduino serial interface. Is there a programming interface  for that? My first idea was to trick SimHub into thinking that a null model (emulated port) is an Arduino board in order to translate serial data into HID reports, but I don't know how SimHub is listing ports and detecting Arduino boards. Thanks in advance.

>> I can't help you fully here, but SimHub detects all Arduino-capable boards first and foremost, regardless of their programming.
It is their code that acknowledges a handshake and allows them to have information send or settings set for them that allows communication.  This is all Serial-traffic based, not HID-based.
The joystick functionality is HID-based, in the subclass of Joysticks, with their own subclass for joystick types like joystick, gamepad and multi-axis.
>> Unfortunately I have not yet worked with the CustomSerial part of SimHub to the degree that it shows up in the "Devices" section, so this is unfortunately where my expertise stops....

> How do commercial sim-wheels (or "rev lights") interface SimHub? Do they implement the serial protocol over USB-CDC? 
>> The ones I had in the G27 used a controller in the base.  The Fanatec wheels use SPI over USB (not sure if there is some USB translation happening in the base, given the fact that there is a USB-C connection in mine ending in a pin-connector meant for the "quick" release, I don't know what is in between the pin header and the USB-C connection so I suspect it is SPI over USB).
>> It's up to the wheel to then translate the information there and light up the LED's.  I take it a part of the way these lights work is by listening to games that support it and checking the RPM directly instead of through SimHub.
>> Given that both need drivers in order to get out of HID baseclasses and become joysticks that are outside the baseclass of regular joysticks if I remember the device information correctly, I suspect there is some driver stuff going on there.
Especially given the fact these lights remained fully on when the drivers were absent.  The Fanatec wheels had the LED's remain off without drivers.  Auto-centering worked, but without drivers there's obviously no force-feedback.

**Reply by afpineda** [reply]
As you can see, this was a waste of time.

**Comment by afpineda** [comment]
I turned to JetBrains' [DotPeek](https://www.jetbrains.com/decompiler/) to **reverse engineer** the SimHub plugin SDK.
I found that the "Devices" and "Arduino" features are just regular plugins. There may be a way to reuse code in order to translate game data into LED color data, including LED profiles.
Don't expect news anytime soon, but it seems promising.

**Reply by janc18** [reply]
That sound interesting to play with it. Do you want to create a new discussion here in GitHub to share findings about how the plugins work.    

Edit:
Perhaps it's better to create a new branch

**Reply by ArtTales** [reply]
Great idea, or as I said, create a discord server, where people interested in this could help eachother and grow project. There would be more conversations like this

**Reply by afpineda** [reply]
> That sound interesting to play with it. Do you want to create a new discussion here in GitHub to share findings about how the plugins work.

Yes, but not here. The SimHub plugin is the place for that. **I already created a new discussion:**
[https://github.com/afpineda/SimWheelESP32-SimHub-Plugin/discussions/1](https://github.com/afpineda/SimWheelESP32-SimHub-Plugin/discussions/1)
> 
> Edit: Perhaps it's better to create a new branch

I'm not extending the code base without a clear understanding about where it leads.
"Taste testing" must go in private repos.

**Reply by afpineda** [reply]
> Great idea, or as I said, create a discord server, where people interested in this could help eachother and grow project. There would be more conversations like this

I will consider your suggestion to build a user community without the need to create a GitHub account.
But I do not have time to moderate a Discord server or read every post.
Don't expect me to participate.

**Reply by janc18** [reply]
Agreed, I'll follow up in the SimHub plugin discussion.

**Comment by afpineda** [comment]
Hi, @ArtTales.

I am pleased to announce that the firmware is entering a new development cycle **to implement this feature**. I have managed to get the [SimHub plugin](https://github.com/afpineda/SimWheelESP32-SimHub-Plugin/tree/development) to work.

**Comment by afpineda** [comment]
This feature was implemented in **version 6.4.0** in conjunction with the *SimHub plugin*.
This was a real pain as I had to resort to reverse engineering SimHub due to a complete lack of documentation.
